peps/pep-0740.rst

```diff
@@ -0,0 +1,288 @@
+
+Rationale
+=========
+
+Desire for digital signatures on Python packages has been repeatedly
+expressed by both package maintainers and downstream users:
+
+* Maintainers wish to demonstrate the integrity and authenticity of their
+  package uploads;
+* Individual downstream users wish to verify package integrity and authenticity
+  without placing additional trust in their index's honesty;
+* "Bulk" downstream users (such as Operating System distributions) wish to
+  perform similar verifications and potentially re-expose or countersign
+  for their own downstream packaging ecosystems.
+
+This proposal seeks to accommodate each of the above use cases.
+
+While this PEP does not recommend a specific digital attestation format,
+it does recognize the utility of Trusted Publishing as a pre-existing,
+"zero-configuration" source of strong provenance for Python packages.
+Consequently this PEP includes a proposed scheme for exposing each release
+file's Trusted Publisher metadata, with the expectation that a future digital
+attestation format will likely make use of it.
+
+Design Considerations
+---------------------
+
+This PEP identifies the following design considerations when evaluating
+both its own proposed changes and previous work in the same or adjacent
+areas of Python packaging:
+
+1. Index accessibility: digital attestations for Python packages
+   are ideally retrievable directly from the index itself, as "detached"
+   resources.
+
+   This both simplifies some compatibility concerns (by avoiding
+   the need to modify the distribution formats themselves) and also simplifies
+   the behavior of potential installing clients (by allowing them to
+   retrieve each attestation before its corresponding package without needing
+   to do streaming decompression).
+
+2. Verification by the index itself: in addition to enabling verification
+   by installing clients, each digital attestation is *ideally* verifiable
+   in some form by the index itself.
+
+   This both increases the overall quality
+   of attestations uploaded to the index (preventing, for example, users
+   from accidentally uploading incorrect or invalid attestations) and also
+   enables UI and UX refinements on the index itself (such as a "provenance"
+   view for each uploaded package).
+
+3. General applicability: digital attestations should be applicable to
+   *any and every* package uploaded to the index, regardless of its format
+   (sdist or wheel) or interior contents.
+
+4. Metadata support: this PEP refers to "digital attestations" rather than
+   just "digital signatures" to emphasize the ideal presence of additional
+   metadata within the cryptographic envelope.
+
+   For example, to prevent domain separation between a distribution's name and
+   its contents, the digital attestation could be performed over
+   ``HASH(name || HASH(contents))`` rather than just ``HASH(contents)``.
+
+5. Consistent release attestations: if a file belonging to a release has a
```

Comment:
Yeah, this is a bit of a wart: the idea behind this is to prevent "mosaic" attestations, which are significantly harder to design UIs/write policies for (and consequently far easier to smuggle un-attested artifacts through).

Whether or not this is a good idea comes down to another question, one that's maybe outside of the scope of this PEP: should PyPI (and other indices) *encourage* the "never done" nature of releases?

My impression is that the "never done" property is *mostly* a historical accident/consequence of the fact that the legacy upload API only accepts a single distribution at a time, and that it has negative repercussions throughout the rest of the ecosystem (e.g. preventing dependency resolution caching/optimizations, since a more specialized wheel can appear at any point).

That, combined with the new (WIP) Upload 2.0 API and the fact that point releases are the "right" way to distribute new assets over a timeline makes me think that preserving this property of releases isn't something that should be a priority. But it's possible that there are scenarios I haven't thought of here, or that there are otherwise good reasons to allow this.

> This restriction also might add friction for tools looking to add new attestations, I believe many Python workflows use different tools to build different artifacts (ie source vs wheel) so those two tools would need to both support an attestation format before being able to upload its attestation.

I might be missing a case, but I think this *shouldn't* add any significant friction in the case of Trusted Publisher-published artifacts or even hypothetical self-held key attested artifacts. Concretely:

* In a Trusted Publisher scenario for package `foo`, `foo.tar.gz` might be published by `pub-sdist.yml` and `foo.whl` might be published by `pub-wheel.yml` (ideally they'd use the same publish workflow, but nothing stops a user from configuring them separately). In this scenario, even though the two file distributions have different "identities" in their publish attestations, the attestation *kind* is invariant (they're both published by a Trusted Publisher and might have a kind of `sigstore:publish`, to make up an arbitrary identifier).
* In a self-held key scenario, the user holds the key and can make any attestation they please with it, at any time.

Does this address the friction concern?

Reply:
> Yeah, this is a bit of a wart: the idea behind this is to prevent "mosaic" attestations, which are significantly harder to design UIs/write policies for (and consequently far easier to smuggle un-attested artifacts through).

Agree with the motivations!

> Whether or not this is a good idea comes down to another question, one that's maybe outside of the scope of this PEP: should PyPI (and other indices) _encourage_ the "never done" nature of releases?

If a new index was being designed today I would not recommend having this property. In practice this property means that PyPI is an edge-case in almost all discussions or statements about software repositories. This is something I have to remind folks outside the Python ecosystem about constantly as a gotcha.

> That, combined with the new (WIP) Upload 2.0 API and the fact that point releases are the "right" way to distribute new assets over a timeline makes me think that preserving this property of releases isn't something that should be a priority. But it's possible that there are scenarios I haven't thought of here, or that there are otherwise good reasons to allow this.

Agree with new releases being the way to distribute new artifacts.

Thinking more, my hesitation might come from the requirement applying to the "release" rather than a single upload action? Using build provenance as example, it's likely that projects will continue to have separate workflows for source/wheel builds and maybe only one of those workflows is capable of generating build provenance. I can see this requirement getting in the way of index (and third-party) attestations as well.

Perhaps this can be a requirement of attestations submitted via the "upload 2.0 API"?

Reply:
> Thinking more, my hesitation might come from the requirement applying to the "release" rather than a single upload action? Using build provenance as example, it's likely that projects will continue to have separate workflows for source/wheel builds and maybe only one of those workflows is capable of generating build provenance. I can see this requirement getting in the way of index (and third-party) attestations as well.
> 
> Perhaps this can be a requirement of attestations submitted via the "upload 2.0 API"?

I think that would work. But to voice a concern: one of the advantages of this work is the afforded opportunity to "get it right" in terms of invariants, meaning that future consumers of whatever we standardize here don't have to worry about mosaics, differentiating between distributions, etc.

Binding to historical quirks here is likely to cause less short-term pain, but (IMO) is likely to cause more pain in the medium- and long-terms as we try to write meaningful security policies over attestations.

Using the build provenance example, I think this actually demonstrates the invariant working well: if a project can only produce provenance for a single artifact but Python dependency resolution doesn't guarantee that artifact's selection (which it can't, due to indeterminate specificity), then IMO it's actually dangerous to allow the project to submit provenance at all (since users will presume that the singular provenance binds all artifacts, or that the others are derivative artifacts, etc.). 

TL;DR: I think we *can* drop this constraint, but IMO we'll rue the decision in the distant future (and will have further trouble explaining the functionality here to users/policy builders, who will need to poke into the guts of Python dep resolution to know whether or not the distribution they're receiving is consistent with others of the same release). But if you think it's too onerous as-is, I'll remove it from the PEP 🙂 

(Taking a step back: maybe we can finally fix this one and for all at the index level? AFAIK the "never done" property isn't standardized at all, so maybe we can change PyPI's policy to "must be uploaded no more than 24 hours after the first file"?)

Reply:
I think it makes sense to want the "all artifacts must provide consistent attestations across one release". You referenced in another chat about retaining the requirement and then later loosening it if the need arises, all of my qualms with it now are for hypothetical futures where we have index or third-party attestations (like build reproducibility) that can't apply across all artifacts. We can revisit this decision when we get to that future.

> (Taking a step back: maybe we can finally fix this one and for all at the index level? AFAIK the "never done" property isn't standardized at all, so maybe we can change PyPI's policy to "must be uploaded no more than 24 hours after the first file"?)

I'm in favor of digging in deeper to how frequently this occurs and potentially limiting new uploads after a set time period. If it were to go forward based on data that this isn't occurring often I could see a campaign similar to username/password for uploads.